File: app/lib/check_update.dart
// GitHub release 检查 + 打开 release 页面 / README 链接
//
// 注意：const String.fromEnvironment 用于 build 时的硬编码值。这里
// APP_VERSION 来自 pubspec.yaml，pub_get 时已注入。

import 'dart:convert';
import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;
import 'package:url_launcher/url_launcher.dart';

class CheckUpdate {
  /// 当前版本（pubspec.yaml 里的 version 字段）
  static const String currentVersion = String.fromEnvironment(
    'APP_VERSION',
    defaultValue: '2.0.0',
  );

  /// 当前 build number
  static const String currentBuild = String.fromEnvironment(
    'APP_BUILD',
    defaultValue: '20',
  );

  /// GitHub 仓库（owner/name）—— v2.0.0 之后写死
  /// 若要换仓库，所有已装 APK 需要重新 build（这就是为啥写死而不是 --dart-define）
  static const String repo = 'mzhscan/monitor-status';

  static String get _apiUrl => 'https://api.github.com/repos/$repo/releases/latest';
  static String get _releasePage => 'https://github.com/$repo/releases/latest';
  static String get _readmeUrl => 'https://github.com/$repo#readme';

  /// Fetch the latest release. Returns null on network / parse error.
  static Future<CheckUpdateResult> fetchLatest() async {
    try {
      final r = await http
          .get(Uri.parse(_apiUrl), headers: {'Accept': 'application/vnd.github+json'})
          .timeout(const Duration(seconds: 10));
      if (r.statusCode != 200) {
        return CheckUpdateResult.error('GitHub 返回 HTTP ${r.statusCode}');
      }
      final j = jsonDecode(r.body) as Map<String, dynamic>;
      final tag = (j['tag_name'] as String?) ?? '';
      final body = (j['body'] as String?) ?? '';
      // Strip leading "v" if present, then semver-compare
      final remote = _parseSemver(tag);
      final local = _parseSemver(currentVersion);
      final isNewer = remote != null && local != null && _compare(remote, local) > 0;
      return CheckUpdateResult(
        latestTag: tag,
        body: body,
        isNewer: isNewer,
      );
    } catch (e) {
      return CheckUpdateResult.error(e.toString());
    }
  }

  static List<int>? _parseSemver(String s) {
    final m = RegExp(r'^v?(\d+)\.(\d+)\.(\d+)').firstMatch(s.trim());
    if (m == null) return null;
    return [int.parse(m.group(1)!), int.parse(m.group(2)!), int.parse(m.group(3)!)];
  }

  /// Returns positive if a > b.
  static int _compare(List<int> a, List<int> b) {
    for (var i = 0; i < 3; i++) {
      if (a[i] != b[i]) return a[i] - b[i];
    }
    return 0;
  }

  static Future<void> openReleasePage(BuildContext context) async {
    final uri = Uri.parse(_releasePage);
    if (await canLaunchUrl(uri)) {
      await launchUrl(uri, mode: LaunchMode.externalApplication);
    }
  }

  static Future<void> openReadme(BuildContext context) async {
    final uri = Uri.parse(_readmeUrl);
    if (await canLaunchUrl(uri)) {
      await launchUrl(uri, mode: LaunchMode.externalApplication);
    }
  }
}

class CheckUpdateResult {
  final String? latestTag;
  final String? body;
  final bool isNewer;
  final bool isError;
  final String? error;

  const CheckUpdateResult({
    this.latestTag,
    this.body,
    this.isNewer = false,
    this.isError = false,
    this.error,
  });

  factory CheckUpdateResult.error(String msg) =>
      CheckUpdateResult(isError: true, error: msg);

  String get summary {
    if (isError) return '检查失败：$error';
    if (latestTag == null || latestTag!.isEmpty) return '无 release 信息';
    if (isNewer) return '有新版本: v$latestTag';
    return '当前已是最新版本 (v$latestTag)';
  }
}

File: ios-native/XingLiMonitor/Core/CheckUpdate.swift
// GitHub release 检查 —— 对齐 Flutter 版 app/lib/check_update.dart。

import Foundation

struct CheckUpdateResult {
    var latestTag: String?
    var body: String?
    var isNewer = false
    var isError = false
    var error: String?

    var summary: String {
        if isError { return "检查失败：\(error ?? "")" }
        guard let tag = latestTag, !tag.isEmpty else { return "无 release 信息" }
        if isNewer { return "有新版本: v\(tag)" }
        return "当前已是最新版本 (v\(tag))"
    }
}

enum CheckUpdate {
    /// GitHub 仓库（owner/name）—— 跟 Dart 版写死同一个
    static let repo = "mzhscan/monitor-status"

    static var releasePage: URL { URL(string: "https://github.com/\(repo)/releases/latest")! }
    static var readmeUrl: URL { URL(string: "https://github.com/\(repo)#readme")! }

    static func currentVersion() -> String {
        Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "3.0.0"
    }

    /// Fetch the latest release. 网络 / 解析错 → isError
    static func fetchLatest() async -> CheckUpdateResult {
        let url = URL(string: "https://api.github.com/repos/\(repo)/releases/latest")!
        var req = URLRequest(url: url)
        req.setValue("application/vnd.github+json", forHTTPHeaderField: "Accept")
        req.timeoutInterval = 10
        do {
            let (data, resp) = try await URLSession.shared.data(for: req)
            guard let http = resp as? HTTPURLResponse else {
                var r = CheckUpdateResult(); r.isError = true; r.error = "响应不是 HTTP"; return r
            }
            if http.statusCode != 200 {
                var r = CheckUpdateResult(); r.isError = true
                r.error = "GitHub 返回 HTTP \(http.statusCode)"; return r
            }
            guard let obj = try? JSONSerialization.jsonObject(with: data),
                  let j = obj as? [String: Any] else {
                var r = CheckUpdateResult(); r.isError = true; r.error = "解析失败"; return r
            }
            let tag = j["tag_name"] as? String ?? ""
            let body = j["body"] as? String ?? ""
            let remote = parseSemver(tag)
            let local = parseSemver(currentVersion())
            var isNewer = false
            if let remote, let local {
                isNewer = compare(remote, local) > 0
            }
            return CheckUpdateResult(latestTag: tag, body: body, isNewer: isNewer)
        } catch {
            var r = CheckUpdateResult()
            r.isError = true
            r.error = "\(error.localizedDescription)"
            return r
        }
    }

    static func parseSemver(_ s: String) -> [Int]? {
        // ^v?(\d+)\.(\d+)\.(\d+)
        let trimmed = s.trimmingCharacters(in: .whitespaces)
        var nums: [Int] = []
        var rest = Substring(trimmed)
        if rest.first == "v" { rest = rest.dropFirst() }
        for _ in 0..<3 {
            var digits = Substring()
            for ch in rest {
                if ch.isNumber { digits.append(ch) } else { break }
            }
            guard !digits.isEmpty, let n = Int(digits) else { return nil }
            nums.append(n)
            rest = rest.dropFirst(digits.count)
            if nums.count < 3 {
                guard rest.first == "." else { return nil }
                rest = rest.dropFirst()
            }
        }
        return nums
    }

    /// a > b 返回正数
    static func compare(_ a: [Int], _ b: [Int]) -> Int {
        for i in 0..<3 where a[i] != b[i] {
            return a[i] - b[i]
        }
        return 0
    }
}
